trinity/protocol/common/managers.py

```diff
@@ -214,3 +237,129 @@ def _send_sub_proto_request(self, request: TRequest) -> None:
 
     def get_stats(self) -> str:
         return '%s: %s' % (self.response_msg_name, self.response_times.get_stats())
+
+
+class RequestManager(PeerSubscriber, BaseService, Generic[TPeer, TMsg]):
+
+    #
+    # PeerSubscriber
+    #
+    @property
+    def subscription_msg_types(self) -> Set[Type[Command]]:
+        return {self._response_msg_type}
+
+    msg_queue_maxsize = 100
+
+    response_timout: int = 60
+
+    pending_request: Tuple[float, 'asyncio.Future[TMsg]'] = None
+
+    _peer: TPeer
+
+    def __init__(self, peer: TPeer, response_msg_type: Type[Command], token: CancelToken) -> None:
+        super().__init__(token)
+        self._peer = peer
+        self.response_times = ResponseTimeTracker()
+        self._response_msg_type = response_msg_type
+
+    async def message_candidates(
+            self,
+            make_request: Callable[[], None],
+            logger: logging.Logger,
+            timeout: int) -> 'AsyncGenerator[TMsg, None]':
+        """
+        Make a request and iterate through candidates for a valid response.
+
+        To mark a response as valid, use the `complete_if_valid` context manager. Within it,
+        validate the received message. If no ValidationError is raised, then the
+        message will be considered delivered, and the response closed.
```

Comment:
I read this twice and I am still not sure what this is telling me to do.  Maybe we could include a code example in the docstring?